packages/itwinui-react/package.json

```diff
@@ -82,7 +82,7 @@
     "classnames": "^2.3.2",
     "jsdom": "^24.0.0",
     "react-table": "^7.8.0",
-    "react-transition-group": "^4.4.5",
+    "react-transition-group": "4.4.2",
```

Comment:
I don't think this is feasible. `react-transition-group` is a very popular library, so our consumers may already be using it, and they might dedupe their deps, causing a different version to be used.

Thinking a bit more about it, we really only need `react-transition-group` for exit animations (specifically, we need to keep the element in DOM for a short duration, then remove it after enimation completes). So we can simplify the React code, and handle everything else using plain CSS.

Reply:
Diff (after reverting your changes) shows what I'm suggesting:

<details>
<summary>dialog.scss</summary>

```diff
@@ -44,12 +44,18 @@ $iui-dialog-min-height: 7.75rem; // 7.75rem = 124px = title bar height + margins
   }
 
   transition:
-    visibility var(--iui-duration-0) linear,
-    opacity var(--iui-duration-1) ease-out;
-  transition-delay: var(--iui-duration-1), var(--iui-duration-0);
+    visibility var(--iui-duration-0) linear var(--iui-duration-1),
+    opacity var(--iui-duration-1) ease-out var(--iui-duration-0);
 
   &.iui-dialog-visible {
-    transition-delay: var(--iui-duration-0); // remove delay for entry animation
+    visibility: visible;
+    animation: iui-fade-in var(--iui-duration-1) ease-out;
+  }
+
+  @keyframes iui-fade-in {
+    from {
+      opacity: 0;
+    }
   }
 }
 
@@ -116,7 +122,13 @@ $iui-dialog-min-height: 7.75rem; // 7.75rem = 124px = title bar height + margins
       transform var(--iui-duration-1) ease-in;
 
     &.iui-dialog-visible {
-      transition: transform var(--iui-duration-2) ease-out;
+      animation: iui-slide-up var(--iui-duration-2) ease-out;
+    }
+  }
+
+  @keyframes iui-slide-up {
+    from {
+      transform: translateY(100%);
     }
   }
 }
@@ -125,18 +137,6 @@ $iui-dialog-min-height: 7.75rem; // 7.75rem = 124px = title bar height + margins
   border: 1px solid var(--iui-color-border);
 }
 
-.iui-dialog-animation {
-  &-enter {
-    transform: translateY(100%);
-    opacity: 0;
-  }
-
-  &-enter-active {
-    transform: translateY(0);
-    opacity: 1;
-  }
-}
-
 .iui-dialog-title {
   display: flex;
   align-items: center;
```
</details>

<details>
<summary>DialogMain.tsx</summary>

```diff
@@ -16,10 +16,9 @@ import {
 import type { PolymorphicForwardRefComponent } from '../utils/index.js';
 import { useDialogContext } from './DialogContext.js';
 import type { DialogContextProps } from './DialogContext.js';
-import { CSSTransition } from 'react-transition-group';
+import { Transition } from 'react-transition-group';
 import { DialogDragContext } from './DialogDragContext.js';
 import { useDragAndDrop } from '../utils/hooks/useDragAndDrop.js';
-import styles from '../../styles.js';
 
 export type DialogMainProps = {
   /**
@@ -169,6 +168,7 @@ export const DialogMain = React.forwardRef((props, ref) => {
         {
           'iui-dialog-default': styleType === 'default',
           'iui-dialog-full-page': styleType === 'fullPage',
+          'iui-dialog-visible': isOpen,
           'iui-dialog-draggable': isDraggable,
         },
         className,
@@ -207,13 +207,8 @@ export const DialogMain = React.forwardRef((props, ref) => {
   );
 
   return (
-    <CSSTransition
+    <Transition
       in={isOpen}
-      classNames={{
-        enter: styles['iui-dialog-animation-enter'],
-        enterActive: styles['iui-dialog-animation-enter-active'],
-        enterDone: styles['iui-dialog-visible'],
-      }}
       timeout={{ exit: 600 }}
       // Focuses dialog when opened
       onEntered={() => {
@@ -238,6 +233,6 @@ export const DialogMain = React.forwardRef((props, ref) => {
         {trapFocus && <FocusTrap>{content}</FocusTrap>}
         {!trapFocus && content}
       </DialogDragContext.Provider>
-    </CSSTransition>
+    </Transition>
   );
 }) as PolymorphicForwardRefComponent<'div', DialogMainProps>;
```
</details>

Reply:
I see. Reverted back `react-transition-group`'s version related changed ([`4642e88` (#1835)](https://github.com/iTwin/iTwinUI/pull/1835/commits/4642e88746155682589bef8e517bd85d3b22f899), https://github.com/iTwin/iTwinUI/pull/1835/commits/20740f7629cf1174c9974db70aa46a5b493a28c3). And used `Transition` instead of `CSSTransition` and doing animations in css like in your diff. (https://github.com/iTwin/iTwinUI/pull/1835/commits/528a10ffbb43b8d72ef49f56cc22e26c4051fe66)

Reply:
Could you also try to fix the html demo? I think you just need to add 2 lines of JS to force the `hidden` attribute when the dialog is closed; this should be enough to trigger the CSS `animation`.

Reply:
I was playing around with adding `hidden` and/or `inert` when the dialog is closed. That did not fix the broken entry animation on the css demo page (https://github.com/iTwin/iTwinUI/pull/1835/commits/cd8201338229f32624c771c1ad3a83167e4be188).

Adding `visibility: hidden;` in the `from` of `iui-fade-in` fixed the entry animation for all dialogs except the full page dialog (https://github.com/iTwin/iTwinUI/pull/1835/commits/b7ca4c806a459f6fffe34468bf3daeaede6097fc). I am still investigating this.

Also, I noticed that when itwinui-react users render the Dialog conditionally, with the latest commits from this PR, we only show the entry animation but not the exit transition/animation.

https://github.com/iTwin/iTwinUI/assets/45748283/0083ae69-dc16-4e8f-ae39-2d357a83f675

So maybe we need to replace the exit transition with an exit animation.

Reply:
> Adding `visibility: hidden;` in the `from` of `iui-fade-in` fixed the entry animation for all dialogs except the full page dialog ([b7ca4c8](https://github.com/iTwin/iTwinUI/commit/b7ca4c806a459f6fffe34468bf3daeaede6097fc)). I am still investigating this.

This was happening because the `iui-fade-in` `animation` attribute from `iui-dialog` was being overridden by the `iui-slide-up` `animation` attribute from `iui-dialog-full-page`.

So, the dialog on the css page was not fading in, but was instead just sliding up whilst being hidden. This is now fixed in [`37e5378` (#1835)](https://github.com/iTwin/iTwinUI/pull/1835/commits/37e537847ea0aae67212ea9ddad313ef5ac1f596).

Reply:
> I was playing around with adding `hidden` and/or `inert` when the dialog is closed. That did not fix the broken entry animation on the css demo page ([cd82013](https://github.com/iTwin/iTwinUI/commit/cd8201338229f32624c771c1ad3a83167e4be188)).

`hidden` attribute's styles are usually too weak, so you need a global style along the lines of:

```css
[hidden] { display: none !important; }
```

It's important that the `display` property changes from `none` → `flex` (or something else) for the animation to work.

`inert` is generally not necessary when `display: none` or `visibility: hidden` is set.

> Or if we're okay with having just an entry animation but no exit animation/transition, then how it is currently is also alright.

I think this is ok. It's expected that the exit animation will break if the user literally does not render the `Dialog`, because the dialog needs to control its own unmounting.


> > Adding `visibility: hidden;` in the `from` of `iui-fade-in` fixed the entry animation for all dialogs except the full page dialog ([b7ca4c8](https://github.com/iTwin/iTwinUI/commit/b7ca4c806a459f6fffe34468bf3daeaede6097fc)). I am still investigating this.
> 
> This was happening because the `iui-fade-in` `animation` attribute from `iui-dialog` was being overridden by the `iui-slide-up` `animation` attribute from `iui-dialog-full-page`.

Ah you might have used an earlier version of my [suggested diff](https://github.com/iTwin/iTwinUI/pull/1835#discussion_r1483645397), which I edited later.

Reply:
> Ah you might have used an earlier version of my [suggested diff](https://github.com/iTwin/iTwinUI/pull/1835#discussion_r1483645397), which I edited later.

I implemented the latest version of your suggestion with minor changes to get the animations working ([`87014d4` (#1835)](https://github.com/iTwin/iTwinUI/pull/1835/commits/87014d4d963d3e42bde182f2341ecb6895334b69), https://github.com/iTwin/iTwinUI/pull/1835/commits/50cbc7f5aa3633fc2f1ce30b4b0492cb6bd46634)

> It's important that the `display` property changes from `none` → `flex` (or something else) for the animation to work.

After reverting all changes in `dialog.html`, the animations still worked. So, I committed this revert (https://github.com/iTwin/iTwinUI/pull/1835/commits/1ccbff8bfadce42584b499936c17fbac75735ea5)

Reply:
> > It's important that the `display` property changes from `none` → `flex` (or something else) for the animation to work.
> 
> After reverting all changes in `dialog.html`, the animations still worked. So, I committed this revert ([1ccbff8](https://github.com/iTwin/iTwinUI/commit/1ccbff8bfadce42584b499936c17fbac75735ea5))

Did you test it in Firefox/Safari? I think Chromium has some newer animation improvements that are not available elsewhere.

Reply:
Ohh, it was not working on Firefox and Safari. But that was due to some CSS issues.

I fixed it by reverting dialog.scss to before https://github.com/iTwin/iTwinUI/pull/1835/commits/87014d4d963d3e42bde182f2341ecb6895334b69 ([revert commit](https://github.com/iTwin/iTwinUI/pull/1835/commits/dde45410652ba0ed671e9532093e9ea4045ebf68)). I then fixed it in [`c99ee3f` (#1835)](https://github.com/iTwin/iTwinUI/pull/1835/commits/c99ee3ff069b114ba455c72ee77adff14ef98f70) and  [`0e96e88` (#1835)](https://github.com/iTwin/iTwinUI/pull/1835/commits/0e96e88c45529085ca2cdfee7334c6777d179cbd).

The fix I used also included changing the iui-fade-in animation duration to equal the iui-slide-up animation duration. Not sure if the durations were initially different for a reason.

I tested that the dialog css demo now works on Chromium and Firefox. It works on Safari too but there's a bit of a delay in the entry animation for the full page dialog. Still looking into this.

<details>
<summary>Screen recordings</summary>

Safari (from this PR):

https://github.com/iTwin/iTwinUI/assets/45748283/a508d0f2-dbae-4c7f-a5d1-193b3ca8c53f

Safari (hosted version):

https://github.com/iTwin/iTwinUI/assets/45748283/c8c706d7-55f5-46c1-afcf-1165e9710fdb

Firefox (from this PR):

https://github.com/iTwin/iTwinUI/assets/45748283/f730f20f-0947-4361-b338-6854d6a74825

</details>

Reply:
It works fine for me in all browsers, including Safari:

https://github.com/iTwin/iTwinUI/assets/9084735/b568ea6f-9836-4c66-8edd-4c731399d07a

Reply:
Great, the entry animation on the hosted version is working on Safari for me too. So, the local issue I'm facing might be unrelated.